### PR TITLE
Precommit formatting

### PR DESCRIPTION
**Pull request by d-v-b**
- adds ruff and black to precommit
- removes flake8 from precommit
- moves codespell configuration to `pyproject.toml` 
- code changes required to appease the linters

* [ ] Add unit tests and/or doctests in docstrings (N/A)
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions (N/A)
* [ ] New/modified features documented in docs/tutorial.rst (N/A)
* [x] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

**Comment by d-v-b** [comment]
commit history looks gnarly -- sorry about that. I accidentally put some commits from a feature branch into my main branch, then reverted at  [99b04b6](https://github.com/zarr-developers/zarr-python/pull/1321/commits/99b04b63899738be9fa754073a3db03b92f2981a)

File: .github/workflows/needs_release_notes.yml [hunk 1]
```diff
@@ -5,7 +5,6 @@ on:
 
 jobs:
   triage:
-    if: ${{ github.event.pull_request.user.login != 'dependabot[bot]' }} && ${{ github.event.pull_request.user.login != 'pre-commit-ci[bot]' }}
```
Comment (joshmoore) [inline .github/workflows/needs_release_notes.yml:8]: Why did this get dropped?

File: .github/workflows/releases.yml [hunk 2]
```diff
@@ -64,7 +64,7 @@ jobs:
         with:
           name: releases
           path: dist
-      - uses: pypa/gh-action-pypi-publish@v1.6.4
```
Comment (joshmoore) [inline .github/workflows/releases.yml:67]: Ah, I think this is reverting things that have hit the mainline.

Comment (d-v-b) [reply .github/workflows/releases.yml:67]: yes, and I'm not sure why this is happening -- possibly we've reached the limits of my experience with git. When I run `git pull upstream main`, git claims everything is up to date, but clearly there's lots of recent commits on mainline that are missing from this PR. Any advice?

**Comment by d-v-b** [comment]
closing in favor of #1323

**Pull request closed without merging**